### PR TITLE
Feature/extend visuals

### PR DESCRIPTION
**Pull request by blondered**
## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Optimization

## How Has This Been Tested?
Before submitting a PR, please check yourself against the following list. It would save us quite a lot of time.
- Have you read the contribution guide?
- Have you updated the relevant docstrings? We're using Numpy format, please double-check yourself
- Does your change require any new tests?
- Have you updated the changelog file?

### CODE REVIEW

File: rectools/visuals/visual_app.py [hunk 1]
```diff
@@ -128,11 +162,15 @@ def __init__(
         auto_display: bool = True,
         formatters: tp.Optional[tp.Dict[str, tp.Callable]] = None,
         rows_limit: int = 20,
+        min_width: int = 100,
         **kwargs: tp.Any,
     ) -> None:
         self.rows_limit = rows_limit
         self.formatters = formatters if formatters is not None else {}
         self.data_storage: _AppDataStorage = self._create_data_storage(*args, **kwargs)
+        if min_width < 10:
```
Comment (feldlime) [inline rectools/visuals/visual_app.py:171]: please, let's not hardcode numbers

At least let's put it into a constant

File: rectools/visuals/visual_app.py [hunk 2]
```diff
@@ -21,14 +23,21 @@ def __init__(
...
-        self.model_names = list(recos.keys())
-        self.request_names = list(selected_requests.keys())
 
+        if n_random_requests > 0:
+            self.selected_requests = self._fill_requests_with_random(
+                selected_requests=selected_requests, n_random_requests=n_random_requests, is_u2i=is_u2i, recos=recos
+            )
+        else:
+            self.selected_requests = selected_requests
+        self.request_names = list(self.selected_requests.keys())
```
[the remark this answers is not in the document]
Comment (blondered) [reply rectools/visuals/visual_app.py]: fixed

File: rectools/visuals/visual_app.py [hunk 3]
```diff
@@ -59,6 +68,31 @@ def __init__(
...
+    @classmethod
+    def _fill_requests_with_random(
+        cls,
+        selected_requests: tp.Dict[tp.Hashable, tp.Hashable],
+        n_random_requests: int,
+        is_u2i: bool,
+        recos: TablesDict,
+    ) -> tp.Dict[tp.Hashable, tp.Hashable]:
+
+        id_col = Columns.User if is_u2i else Columns.TargetItem
```
[the remark this answers is not in the document]
Comment (blondered) [reply rectools/visuals/visual_app.py]: fixed

File: rectools/visuals/visual_app.py [hunk 4]
```diff
@@ -59,6 +68,31 @@ def __init__(
...
+
+        # leave only those ids that were not predefined by user
```
[the remark this answers is not in the document]
Comment (blondered) [reply rectools/visuals/visual_app.py]: fixed

File: rectools/visuals/visual_app.py [hunk 5]
```diff
@@ -59,6 +68,31 @@ def __init__(
...
+        # request ids (e.g. user ids) are stored as values in `selected_requests`
+        all_ids: pd.Series = pd.concat([model_recos[id_col] for model_recos in recos.values()])
+        all_ids = all_ids.drop_duplicates()
+        selected_ids = np.array(list((selected_requests.values())))
```
[the remark this answers is not in the document]
Comment (blondered) [reply rectools/visuals/visual_app.py]: fixed

File: rectools/visuals/visual_app.py [hunk 6]
```diff
@@ -59,6 +68,31 @@ def __init__(
...
+        selected_mask = fast_isin(all_ids.values, selected_ids)
+        selecting_from = all_ids[~selected_mask]
+
+        num_selecting = min(len(selecting_from), n_random_requests)
+        new_ids = np.random.choice(selecting_from, num_selecting, replace=False)
+        upd: tp.Dict[tp.Hashable, tp.Hashable] = {f"random_{i+1}": new_id for i, new_id in enumerate(new_ids)}
```
[the remark this answers is not in the document]
Comment (blondered) [reply rectools/visuals/visual_app.py]: fixed

File: rectools/visuals/visual_app.py [hunk 7]
```diff
@@ -59,6 +68,31 @@ def __init__(
...
+        n_random_requests: int,
+        is_u2i: bool,
+        recos: TablesDict,
+    ) -> tp.Dict[tp.Hashable, tp.Hashable]:
+
+        id_col = Columns.User if is_u2i else Columns.TargetItem
+
+        # leave only those ids that were not predefined by user
+        # request ids (e.g. user ids) are stored as values in `selected_requests`
+        all_ids: pd.Series = pd.concat([model_recos[id_col] for model_recos in recos.values()])
```
[the remark this answers is not in the document]
Comment (blondered) [reply rectools/visuals/visual_app.py]: fixed

File: rectools/visuals/visual_app.py (hunk 1 above)
Comment (blondered) [reply rectools/visuals/visual_app.py]: fixed

File: rectools/visuals/visual_app.py [hunk 8]
```diff
@@ -49,59 +61,91 @@ def __init__(
...
+        selected_requests: tp.Dict[tp.Hashable, tp.Hashable],
+        n_random_requests: int,
+        id_col: str,
+        recos: TablesDict,
+    ) -> tp.Dict[tp.Hashable, tp.Hashable]:
+
+        # Leave only those ids that were not predefined by user
+        # Request ids (e.g. user ids) are stored as values in `selected_requests`
+        all_ids = [model_recos[id_col].unique() for model_recos in recos.values()]
+        unique_ids = np.unique(np.hstack(all_ids))
```
Comment (feldlime) [inline rectools/visuals/visual_app.py:94]: usually it's faster to use `pd.unique`
image

Comment (blondered) [reply rectools/visuals/visual_app.py]: something pandas is faster then something numpy, now that's a finding :)
fixed

Comment (feldlime) [reply rectools/visuals/visual_app.py]: haha

pandas is using a hash table here and numpy is using sorting, so unless you need sorted unique values, pandas is faster by design

**Review by feldlime — APPROVED** [review]

**Pull request merged**